crates/bevy_ecs/src/schedule/ambiguity_detection.rs

```diff
@@ -0,0 +1,284 @@
+                }
+            }
+        }
+        processed.insert(index_a);
+    }
+    ambiguities
+}
+
+impl ReportExecutionOrderAmbiguities {
+    /// Disables all messages reported by the ambiguity checker.
+    pub fn off() -> Self {
+        Self {
+            level: AmbiguityReportLevel::Off,
+            ..Default::default()
+        }
+    }
+
+    /// Displays only the number of unresolved ambiguities detected by the ambiguity checker. This is the default behavior.
+    pub fn minimal() -> Self {
+        Self {
+            level: AmbiguityReportLevel::Minimal,
+            ..Default::default()
+        }
+    }
+
+    /// Displays a full report of ambiguities detected by the ambiguity checker.
+    pub fn verbose() -> Self {
+        Self {
+            level: AmbiguityReportLevel::Verbose,
+            ..Default::default()
+        }
+    }
+
+    /// Adds the given crate to be ignored by ambiguity checker. Check [`ReportExecutionOrderAmbiguities`] for more details.
+    pub fn ignore(mut self, crate_prefix: &str) -> Self {
+        self.ignore_crates.push(crate_prefix.to_string());
+        self
+    }
+
+    /// Adds all the given crates to be ignored by ambiguity checker. Check [`ReportExecutionOrderAmbiguities`] for more details.
+    pub fn ignore_all(mut self, crate_prefixes: &[&str]) -> Self {
+        for s in crate_prefixes {
+            self.ignore_crates.push(s.to_string());
+        }
+        self
+    }
+}
+
+impl Default for ReportExecutionOrderAmbiguities {
+    fn default() -> Self {
+        Self {
+            level: AmbiguityReportLevel::Minimal,
+            ignore_crates: vec![],
+        }
+    }
+}
+
+impl SystemStage {
+    /// Logs execution order ambiguities between systems. System orders must be fresh.
+    pub fn report_ambiguities(&self, world: &mut World) {
+        let ambiguity_report =
+            world.get_resource_or_insert_with(ReportExecutionOrderAmbiguities::default);
+
+        if ambiguity_report.level == AmbiguityReportLevel::Off {
+            return;
+        }
+
+        debug_assert!(!self.systems_modified);
+
+        fn write_display_names_of_pairs(
+            offset: usize,
+            systems: &[impl SystemContainer],
+            ambiguities: Vec<(usize, usize, Vec<ComponentId>)>,
+            world: &World,
+        ) -> usize {
+            for (i, (system_a_index, system_b_index, conflicting_indexes)) in
+                ambiguities.iter().enumerate()
+            {
+                let _system_a_name = systems[*system_a_index].name();
+                let _system_b_name = systems[*system_b_index].name();
+
+                let _conflicting_components = conflicting_indexes
+                    .iter()
+                    .map(|id| world.components().get_info(*id).unwrap().name())
+                    .collect::<Vec<_>>();
+
+                let _ambiguity_number = i + offset;
+
+                println!(
```

Comment:
Is there a reason we don't use `info!` or `warn!` in place of these `println!` calls?

Reply:
info and warn have boilerplate at the start of each message. I think it's probably correct, but I'll need to concatenate the messages before returning them.

Reply:
I tried fixing this, and it results in a lot of spam: 

![image](https://user-images.githubusercontent.com/3579909/160474757-4114c3c5-c7c2-415a-bcfa-66c1fca378cb.png)

I don't think this is feasible to do in this PR.

Reply:
Hm, yeah probably not worth it for now then.

Reply:
Can you collect all the output into one string and just use one `warn!`?

Reply:
I can, but it'll still have the `bevy_ecs::schedule::ambiguity_detection` header. Probably worth though.